### PR TITLE
[system-probe] Reorg ebpf code for reuse with CO-RE network tracer

### PR DESCRIPTION
**Pull request by hmahmood**
### What does this PR do?

Reorganizes the network tracer ebpf code to share implementations of kprobes with the (currently) three versions of the network tracer: prebuilt, runtime compiled, and fentry. This is in preparation for adding a new variant, CO-RE, for the network tracer.

### Motivation

Currently we have copies of the same code in different .c files for prebuilt, rc and fentry. This PR attempts to move all the shared/copied code to one place. This will avoid having to create yet another copy of the code for the co-re tracer.

### Reviewer's Checklist

- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [x] Changed code has automated tests for its functionality.
- [x] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [x] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

**Review by guyarb — APPROVED** [review]

File: pkg/network/ebpf/c/prebuilt/tracer.c [hunk 1]
```diff
@@ -54,596 +31,6 @@ int socket__classifier_dbs(struct __sk_buff *skb) {
...
-    }
-
-    struct sock *sk = (struct sock *)PT_REGS_PARM1(ctx);
-    struct sk_buff *skb = (struct sk_buff *)PT_REGS_PARM2(ctx);
-    int len = (int)PT_REGS_PARM3(ctx);
-    handle_skb_consume_udp(sk, skb, len);
-    return 0;
-}
-
 SEC("kprobe/tcp_retransmit_skb")
```
Comment (akarpz) [inline pkg/network/ebpf/c/prebuilt/tracer.c:34]: why was this left here?

Comment (hmahmood) [reply pkg/network/ebpf/c/prebuilt/tracer.c:34]: We have a different implementation for prebuilt than runtime compiled or co-re for this particular kprobe.

File: pkg/network/ebpf/c/ipv6.h [hunk 2]
```diff
@@ -25,13 +29,8 @@ __maybe_unused static __always_inline bool is_ipv4_mapped_ipv6(__u64 saddr_h, __
...
 static __always_inline void read_in6_addr(u64 *addr_h, u64 *addr_l, const struct in6_addr *in6) {
-#ifdef COMPILE_PREBUILT
-    bpf_probe_read_kernel_with_telemetry(addr_h, sizeof(u64), (void *)&(in6->in6_u.u6_addr32[0]));
-    bpf_probe_read_kernel_with_telemetry(addr_l, sizeof(u64), (void *)&(in6->in6_u.u6_addr32[2]));
-#else
-    BPF_CORE_READ_INTO(addr_h, in6, in6_u.u6_addr32[0]);
-    BPF_CORE_READ_INTO(addr_l, in6, in6_u.u6_addr32[2]);
-#endif
+    *addr_h = *(u64*)(&in6->in6_u.u6_addr32[0]);
+    *addr_l = *(u64*)(&in6->in6_u.u6_addr32[2]);
```
Comment (leeavital) [inline pkg/network/ebpf/c/ipv6.h:33]: Why is it safe to get rid of the prebuilt version here? Is in6_addr a stable struct ?

File: pkg/network/ebpf/c/netns.h [hunk 3]
```diff
@@ -16,36 +14,47 @@
...
+    return net_ns_inum;
+}
+
+#elif defined(COMPILE_CORE) || defined(COMPILE_RUNTIME)
+
+static __always_inline u32 get_netns_ino(struct net* ns) {
     u32 net_ns_inum = 0;
-#ifdef CONFIG_NET_NS
+#if !defined(COMPILE_RUNTIME) || defined(CONFIG_NET_NS)
 #if defined(_LINUX_NS_COMMON_H) || defined(COMPILE_CORE)
```
Comment (leeavital) [inline pkg/network/ebpf/c/netns.h:32]: Should we define a verson of get_netns_ino for RC and CO-RE separately? This level of if statements is very hard to follow.

File: pkg/network/ebpf/c/netns.h [hunk 4]
```diff
@@ -16,36 +14,47 @@
...
     return net_ns_inum;
 }
 
 static __maybe_unused __always_inline u32 get_netns_from_sock(struct sock *sk) {
-    struct net *ct_net = NULL;
-#ifdef CONFIG_NET_NS
-    log_debug("get_netns_from_sock\n");
-    BPF_CORE_READ_INTO(&ct_net, sk, sk_net);
+    struct net *ns = NULL;
+#if !defined(COMPILE_RUNTIME) || defined(CONFIG_NET_NS)
```
Comment (leeavital) [inline pkg/network/ebpf/c/netns.h:43]: Does this actually get hit ? I thought we weren't using CORE with CONFIG_NET_NS ?

File: pkg/network/ebpf/c/sock.h [hunk 5]
```diff
@@ -41,50 +42,146 @@ static __always_inline struct inet_sock *inet_sk(const struct sock *sk)
...
+#elif defined(COMPILE_CORE) || defined(COMPILE_RUNTIME)
+    BPF_CORE_READ_INTO(&sk, sock, sk);
+#endif
+    return sk;
+}
+
 static __always_inline void get_tcp_segment_counts(struct sock* skp, __u32* packets_in, __u32* packets_out) {
+    if (!packets_in || !packets_out) {
+        return;
+    }
```
Comment (leeavital) [inline pkg/network/ebpf/c/sock.h:72]: Why did this get added?

File: pkg/network/ebpf/c/tracer-bind.h [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+#include "bpf_telemetry.h"
+#include "bpf_helpers_custom.h"
+
+#include "tracer.h"
+#include "port.h"
+#include "sock.h"
+#include "tracer.h"
+#include "tracer-maps.h"
+
+//region sys_enter_bind
```
Comment (leeavital) [inline pkg/network/ebpf/c/tracer-bind.h:20]: Is this comment significant ?

**Review by leeavital — COMMENTED** [review]
I don't know if it's easy to do with git, but this commit might be a little easier to read if the break up of pkg/network/ebpf/c/tracer-stats.h wasn't included. Since it's split between several files, it looks like a lot of net-new code and it's not easy to see what I should review.

File: pkg/network/ebpf/c/ipv6.h (hunk 2 above)
Comment (hmahmood) [reply pkg/network/ebpf/c/ipv6.h:33]: Actually this doesn't seem to work always, I'll revert this change.

File: pkg/network/ebpf/c/netns.h (hunk 4 above)
Comment (hmahmood) [reply pkg/network/ebpf/c/netns.h:43]: Yes, this will go through for prebuilt and co-re without checking for CONFIG_NET_NS. For runtime, it will check CONFIG_NET_NS.

File: pkg/network/ebpf/c/sock.h (hunk 5 above)
Comment (hmahmood) [reply pkg/network/ebpf/c/sock.h:72]: I don't think it has any affect. I think I was debugging something and thought this might be the cause. But no harm keeping it there.

Comment (leeavital) [reply pkg/network/ebpf/c/sock.h:72]: Without knowing more I'd prefer to remove it. Having code that we expect to never get hit makes reading the code harder. 

Also: from a more functional perspective, doesn't this let the programs that pass null into get_tcp_segment_counts() without triggering the verifier?   This change might make it easier to make a coding 
mistake elsewhere.

File: pkg/network/ebpf/c/tracer-bind.h (hunk 6 above)
Comment (hmahmood) [reply pkg/network/ebpf/c/tracer-bind.h]: It doesn't seem like it; I am going to remove it.

File: pkg/network/ebpf/c/protocols/classification/maps.h [hunk 7]
```diff
@@ -7,7 +7,7 @@
 #include "protocols/classification/structs.h"
 
 // Kernels before 4.7 do not know about per-cpu array maps.
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 7, 0)
+#if defined(COMPILE_CORE) || LINUX_VERSION_CODE >= KERNEL_VERSION(4, 7, 0)
```
Comment (brycekahle) [inline pkg/network/ebpf/c/protocols/classification/maps.h:10]:
```suggestion
#if defined(COMPILE_CORE) || (COMPILE_RUNTIME && LINUX_VERSION_CODE >= KERNEL_VERSION(4, 7, 0))
```
Otherwise prebuilt will use per-cpu since it gets built on 4.7+

File: pkg/network/tracer/tracer_test.go [hunk 8]
```diff
@@ -976,21 +976,23 @@ func (s *UDPServer) Run(payloadSize int) error {
...
 	}
 
+	s.ln = ln
 	s.address = s.ln.LocalAddr().String()
 
 	go func() {
 		buf := make([]byte, payloadSize)
 		for {
-			n, addr, err := s.ln.ReadFrom(buf)
+			n, addr, err := ln.ReadFrom(buf)
```
Comment (brycekahle) [inline pkg/network/tracer/tracer_test.go:995]: Just curious what motivated the changes in this file? They seem to be functionally identical.

File: pkg/network/ebpf/c/tcp-recv.h [hunk 9]
```diff
@@ -7,46 +7,56 @@
 #include "tracer-maps.h"
 #include "sock.h"
 
-int __always_inline handle_tcp_recv(u64 pid_tgid, struct sock *skp, int recv) {
-    conn_tuple_t t = {};
-    if (!read_conn_tuple(&t, skp, pid_tgid, CONN_TYPE_TCP)) {
+SEC("kprobe/tcp_recvmsg")
+int kprobe__tcp_recvmsg(struct pt_regs *ctx) {
```
Comment (brycekahle) [inline pkg/network/ebpf/c/tcp-recv.h:11]: Is this function only for RC?

File: pkg/network/ebpf/c/tcp-recv.h [hunk 10]
```diff
@@ -7,46 +7,56 @@
...
+    u64 pid_tgid = bpf_get_current_pid_tgid();
+#if defined(COMPILE_RUNTIME) && LINUX_VERSION_CODE < KERNEL_VERSION(4, 1, 0)
+    struct sock *skp = (void*)PT_REGS_PARM2(ctx);
+    int flags = (int)PT_REGS_PARM6(ctx);
+#elif defined(COMPILE_RUNTIME) && LINUX_VERSION_CODE < KERNEL_VERSION(5, 19, 0)
+    struct sock *skp = (struct sock *)PT_REGS_PARM1(ctx);
+    int flags = (int)PT_REGS_PARM5(ctx);
+#else
+    struct sock *skp = (void*)PT_REGS_PARM1(ctx);
```
Comment (brycekahle) [inline pkg/network/ebpf/c/tcp-recv.h:20]:
```suggestion
    struct sock *skp = (struct sock*)PT_REGS_PARM1(ctx);
```

File: pkg/network/ebpf/c/tcp-recv.h [hunk 11]
```diff
@@ -7,46 +7,56 @@
...
 #include "sock.h"
 
-int __always_inline handle_tcp_recv(u64 pid_tgid, struct sock *skp, int recv) {
-    conn_tuple_t t = {};
-    if (!read_conn_tuple(&t, skp, pid_tgid, CONN_TYPE_TCP)) {
+SEC("kprobe/tcp_recvmsg")
+int kprobe__tcp_recvmsg(struct pt_regs *ctx) {
+    u64 pid_tgid = bpf_get_current_pid_tgid();
+#if defined(COMPILE_RUNTIME) && LINUX_VERSION_CODE < KERNEL_VERSION(4, 1, 0)
+    struct sock *skp = (void*)PT_REGS_PARM2(ctx);
```
Comment (brycekahle) [inline pkg/network/ebpf/c/tcp-recv.h:14]:
```suggestion
    struct sock *skp = (struct sock *)PT_REGS_PARM2(ctx);
```

File: pkg/network/ebpf/c/tcp-recv.h [hunk 12]
```diff
@@ -7,46 +7,56 @@
...
+SEC("kprobe/tcp_recvmsg")
+int kprobe__tcp_recvmsg__pre_4_1_0(struct pt_regs* ctx) {
+    u64 pid_tgid = bpf_get_current_pid_tgid();
+    log_debug("kprobe/tcp_recvmsg: pid_tgid: %d\n", pid_tgid);
+    int flags = (int)PT_REGS_PARM6(ctx);
     if (flags & MSG_PEEK) {
         return 0;
     }
 
+    struct sock *skp = (void*)PT_REGS_PARM2(ctx);
```
Comment (brycekahle) [inline pkg/network/ebpf/c/tcp-recv.h:54]:
```suggestion
    struct sock *skp = (struct sock *)PT_REGS_PARM2(ctx);
```

File: pkg/network/ebpf/c/tracer-stats.h [hunk 13]
```diff
@@ -258,32 +262,46 @@ static __always_inline void handle_tcp_stats(conn_tuple_t* t, struct sock* sk, u
     update_tcp_stats(t, stats);
 }
 
-#if defined(COMPILE_RUNTIME) || defined(COMPILE_PREBUILT)
-
-static __always_inline int sk_buff_to_tuple(struct sk_buff *skb, conn_tuple_t *t);
+static __always_inline int handle_skb_consume_udp(struct sock *sk, struct sk_buff *skb, int len) {
+    u64 pid_tgid = bpf_get_current_pid_tgid();
+    udp_recv_sock_t *st = bpf_map_lookup_elem(&udp_recv_sock, &pid_tgid);
+    if (!st) { // no entry means a peek
```
Comment (brycekahle) [inline pkg/network/ebpf/c/tracer-stats.h:268]: I think the `len` check should come first, because it is cheaper than a map lookup.

File: pkg/network/ebpf/c/tracer-stats.h [hunk 14]
```diff
@@ -258,32 +262,46 @@ static __always_inline void handle_tcp_stats(conn_tuple_t* t, struct sock* sk, u
...
     log_debug("skb_consume_udp: bytes=%d\n", data_len);
-    u64 pid_tgid = bpf_get_current_pid_tgid();
     t.pid = pid_tgid >> 32;
     t.netns = get_netns_from_sock(sk);
-    handle_message(&t, 0, data_len, CONN_DIRECTION_UNKNOWN, 0, 1, PACKET_COUNT_INCREMENT, sk);
+    return handle_message(&t, 0, data_len, CONN_DIRECTION_UNKNOWN, 0, 1, PACKET_COUNT_INCREMENT, sk);
 }
 
-#endif // defined(COMPILE_RUNTIME) || defined(COMPILE_PREBUILT)
+int __always_inline handle_tcp_recv(u64 pid_tgid, struct sock *skp, int recv) {
```
Comment (brycekahle) [inline pkg/network/ebpf/c/tracer-stats.h:292]:
```suggestion
static __always_inline int handle_tcp_recv(u64 pid_tgid, struct sock *skp, int recv) {
```

**Review by brycekahle — CHANGES_REQUESTED** [review]
minor changes requested

File: pkg/network/ebpf/c/protocols/classification/maps.h (hunk 7 above)
Comment (hmahmood) [reply pkg/network/ebpf/c/protocols/classification/maps.h:10]: This does get edited here: https://github.com/datadog/datadog-agent/blob/hasan.mahmood/co-re/pkg/network/tracer/connection/kprobe/tracer.go#L164

... but I'll make this change anyway.

File: pkg/network/tracer/tracer_test.go (hunk 8 above)
Comment (hmahmood) [reply pkg/network/tracer/tracer_test.go:995]: `s.ln` is set to `nil` in `UDPServer.Shutdown`; this was causing a crash here.

File: pkg/network/ebpf/c/protocols/classification/maps.h (hunk 7 above)
Comment (brycekahle) [reply pkg/network/ebpf/c/protocols/classification/maps.h]: Hmm, does it make sense to have it always be per-cpu in source then? Since it will get edited if not supported?

**Review by akarpz — APPROVED** [review]
lgtm... I think if you rebase/merge again the macos CI check will start passing but it's not relevant anyway

**Review by brycekahle — APPROVED** [review]
